app/code/Magento/Catalog/Model/ResourceModel/Product/CategoryLink.php

```diff
@@ -115,10 +115,14 @@ private function processCategoryLinks($newCategoryPositions, &$oldCategoryPositi
     {
         $result = ['changed' => [], 'updated' => []];
         foreach ($newCategoryPositions as $newCategoryPosition) {
-            $key = array_search(
-                $newCategoryPosition['category_id'],
-                array_column($oldCategoryPositions, 'category_id')
-            );
+            $key = false;
+
+            foreach ($oldCategoryPositions as $oldKey => $oldCategoryPosition) {
```

Comment:
Hi @kanduvisla. This point is not clear enough for me. 
As far as I see, the new approach based on the loop reflects absolutely the same logic that was previously based on `array_search`. Is the difference in the strict types comparison (if yes, we could use `strict` parameter in `array_search`)? 
I might miss something, so any additional information is appreciated. 
Thank you.

Reply:
Hi @rogyar ,

The problem with the previous implementation is not the fact that `array_search()` was used, but that `array_column()` was used to get the `$key` of `$oldCategoryPositions`. `array_column()` recreates the index, so the second time the loop was run, it could possibly return an index of `0`, which could already have been `unset()` on `$oldCategoryPositions()`, triggering an undefined index.

This bug can be exposed by running the extra testcase I added to the unit test:

    [
        [
            ['category_id' => 3, 'position' => 10],
            ['category_id' => 4, 'position' => 20],
        ],
        [
            ['category_id' => 3, 'position' => 20], // swapped positions
            ['category_id' => 4, 'position' => 10], // swapped positions
        ],
        [
            'update' => [3, 4],
            'insert' => [],
            'delete' => [],
        ],
    ]

The above testcase is perfectly valid, because it is a simple swap of position, so it should trigger 2 updates. On the first run (in `$insertUpdate = $this->processCategoryLinks($categoryLinks, $oldCategoryLinks);`), the index of `category_id => 3` on `$oldCategoryPositions` is `0`. Coincidentally, this is also the same index returned by the `array_column()`-function. So `$key = 0` and later on `unset($oldCategoryPositions[$key]);` is done.

Now comes the bug: on the second run (in `$deleteUpdate = $this->processCategoryLinks($oldCategoryLinks, $categoryLinks);`), the `array_column()` would still return `0` (because of the `unset()` earlier). However, since we've already `unset()` this one, the `unset($oldCategoryPositions[$key]);` will now throw a notice, which in turn will trigger an exception further on.

Even if you don't merge this PR, you can replicate this by adding the extra testcase to the unit test, and replace the last line:

        [
            'update' => [3, 4],
            'insert' => [],
            'delete' => [],
        ],

to:

    [3,4]

But I also updated the unit test to see if the proper CRUD operations are done.

Now as for your question: I don't know if `array_search()` can still be used since we're dealing with a multidimensional array. Perhaps some other `array_*`-function could be used with something like a callback, but because the operation is quite a simple check and we do want to maintain our index I choose for a `foreach` in a `foreach` (forgive me lord for I have sinned).

Reply:
Hi @rogyar ,

Any update on this?